✨ Add Dagoma D6 as found in DiscoUltimate v2 TMC

Sophist-UK opened the pull request:
### Description

Add the Dagoma3D D6 motherboard as used in their Disco Ultimate v2 TMC printer (and possibly other models).

![image](https://reprap.org/forum/thumbcache/537/54e/30a/481/a6b/f75/88f/514/674/a4f/d4_800x400.png)

### Requirements

Dagom3D D6 motherboard

### Benefits

Adds support for a new motherboard. 

### Configurations

Taken from the [Dagoma fork of Marlin DU_MC branch](https://github.com/dagoma3d/Marlin/blob/DU_MC/Marlin/Configuration.h) (where it is called FYSETC_DAGOMA_F5) and explicitly confirmed by Dagoma as being the D6 in response to a ticket I raised with them:

"the BOARD_FYSETC_DAGOMA_F5 is effectively the definition for the D6"

### Related Issues

No.

Sophist-UK committed:
Add Dagoma D6 board as used in their DiscoUltimate v2 TMC.

Taken from the Dagoma fork of Marlin DU_MC branch where it is called FYSETC_DAGOMA_F5 and explicitly confirmed by Dagoma as being the D6:

"the BOARD_FYSETC_DAGOMA_F5 is effectively the definition for the D6"

Sophist-UK committed:
Update pins_DAGOMA_D6.h

ellensp commented:
env:mega2560 does not support io pins > 69

On This board [XYZE0]_SERIAL_RX_PIN and [XYZE0]_SERIAL_TX_PIN are set to IO pin 73.

for IO pins 70-85 you must used env:mega2560ext

Sophist-UK commented:
My mistake - I failed to copy the Dagoma comments. Pushing a fix now.

Sophist-UK committed:
Fix pins.h env for Dagoma D6

InsanityAutomation commented:
A dev working for Dagoma used to be active on discord a few years back. I sent him a ping to see if he was still active and/or employed there to chime in if there is anything to note.

InsanityAutomation commented on Marlin/src/pins/pins.h:
```diff
@@ -224,6 +224,8 @@
 #elif MB(DAGOMA_F5)
   #include "ramps/pins_DAGOMA_F5.h"                 // ATmega2560                           env:mega2560
+#elif MB(DAGOMA_D6)
+  #include "ramps/pins_DAGOMA_D6.h"                 // ATmega2560                           env:FYSETC_F6
```
This should be env:mega2560ext instead of FYSETC_F6

Sophist-UK replied:
FYSETC_F6 was what Dagoma used. But I will change this once I am up and moving.

InsanityAutomation commented:
Orel confirmed on discord that the naming was retained from a prototype board and everything matches, so just the env call to adjust.

Sophist-UK committed:
Address review comments.

0r31 commented:
@ellensp : The board is based on the Fysetc F6, that's whay we use pin 73.

@Sophist-UK : The right env to set is `env:FYSETC_F6`.

(Thanks @InsanityAutomation to ping me ;))

0r31 commented:
Short explanation : 
From platformio platform-atmelavr : https://github.com/platformio/platform-atmelavr/blob/d9d2738bcb1c632e2f5e3c57318e6ae8281ee960/boards/fysetc_f6_13.json#L13
There is a **fysetcf6** variant.

I don't exactly know from where the pin definition file is retrieved (i don't find it https://github.com/arduino/ArduinoCore-avr), but in the platformio framework-arduino-avr package on my computer, the variant file exists : `variants\fysetcf6\pins_arduino.h`.

Inside this file, the pin 73 is explicity defined and used : 
![image](https://github.com/MarlinFirmware/Marlin/assets/37673727/73a6d3f7-8c6b-44ab-b775-f3825bad69f0)

InsanityAutomation commented:
@0r31 It appears at a glance to be a limited implementation of what mega2560ext has, as it likely predates it, with just what they needed. Id like to see if it behaves with the standard definition in order to avoid confusion calling for a fysetc environment with a board not labeled fysetc, but if we cant confirm that then I suppose leaving it as-is works without any additional effort.

0r31 commented:
@InsanityAutomation understood. I didn't know the fysetcf6 variant is "derived" from mega2560ext.

Sophist-UK committed:
Revert "Address review comments."

This reverts commit e2ed70cff97a9d8dd745ab2ba64634a7d37d5813.

thisiskeithb commented:
@Sophist-UK: Please test this board with the `mega2560ext` environment and let us know if it still works as expected.

InsanityAutomation commented:
1 more hindsight thought here... The TMC slave addresses are sometimes in configs sometimes in pins... Should we move this one to the pins file to simplify things for end users?

Sophist-UK commented:
@thisiskeithb Sorry, but I don't have either the D6 board or the printer it is used in. I am trying to generate up to date versions of firmware for a variety of models equivalent to the out-of-date builds provided by the manufacturer. 

But this is the board definition taken verbatim from the Marlin fork that they built their own firmware with.

Sophist-UK commented:
> The TMC slave addresses are sometimes in configs sometimes in pins... Should we move this one to the pins file to simplify things for end users?

I don't know what this means - where would I find these TMC Slave Addresses defined in the Dagoma fork of Marlin for this board? In `config_adv.h`?

thisiskeithb commented:
> 1 more hindsight thought here... The TMC slave addresses are sometimes in configs sometimes in pins... Should we move this one to the pins file to simplify things for end users?

We can assert them in the pins files like I did in https://github.com/MarlinFirmware/Marlin/pull/26842 so configs don’t get recycled into a new build with a different motherboard.

thisiskeithb commented:
> But this is the board definition taken verbatim from the Marlin fork that they built their own firmware with.

Right, but we’re trying to generalize it so it’s not confusing for anyone looking at this in the future.

0r31 commented:
@Sophist-UK @InsanityAutomation : The slave addresses are defined in the config file, as Marlins states it should be.

https://github.com/dagoma3d/Marlin/blob/67da3d0e8c1abb037fc86677953e36fa045704e2/Marlin/Configuration_adv.h#L2947-L2965

But to be printer agnostic, maybe it should be a good idea to add it in the pin definition file. I don't know which is the right convention to follow.

Sophist-UK committed:
Define TMC slave addresses in the pins file as requested

Sophist-UK commented:
I hope that I have done as requested by merging what @or31 pointed in in the format that @thisiskeithb pointed to.

0r31 commented:
@Sophist-UK : seems good to me.

I will test the board with the mega2560ext env before the end of the week. If it works, i will notify you then you could go to this env.

Sophist-UK commented:
> I will test the board with the mega2560ext env before the end of the week.

Thank you - that is extremely good of you to offer.

0r31 commented:
Hi,

Just a quick update. Today i shortly checked :
- The fmk arduino pin file related to mega2560ext ---> ok
- The firmware compilation using this env ---> ok

Tomorrow ---> Tests

0r31 commented:
@Sophist-UK,

It is a final go for me to use the `mega2560ext` env. The printer i use to test behaves exactly the same as before.

Thanks @InsanityAutomation and @thisiskeithb for your wise advices.

Sophist-UK committed:
Reapply "env:mega2560ext"

This reapplies the same fix suggested by early review comments.

Thanks to @0r31 for testing this. :thumbsup:

Sophist-UK commented:
I have reapplies the `env:mega2560ext` fix suggested by early review comments.

Thanks to @0r31 for testing this. 👍

thisiskeithb commented on Marlin/src/core/boards.h:
```diff
@@ -131,6 +131,7 @@
 #define BOARD_PXMALION_CORE_I3        1164  // Pxmalion Core I3
 #define BOARD_PANOWIN_CUTLASS         1165  // Panowin Cutlass (as found in the Panowin F1)
 #define BOARD_KODAMA_BARDO            1166  // Kodama Bardo V1.x (as found in the Kodama Trinus)
+#define BOARD_DAGOMA_D6               1167  // Dagoma D6
```
It'd be helpful to note which machine this board is from like the comments above it. Like so: 

`// Dagoma D6 (as found in the DiscoUltimate V2 TMC)`

thisiskeithb commented on Marlin/src/pins/ramps/pins_DAGOMA_D6.h:
```diff
@@ -0,0 +1,110 @@
+/**
+ * Marlin 3D Printer Firmware
+ * Copyright (c) 2020 MarlinFirmware [https://github.com/MarlinFirmware/Marlin]
```
This should be `(c) 2024`

thisiskeithb commented on Marlin/src/pins/ramps/pins_DAGOMA_D6.h:
```diff
@@ -0,0 +1,110 @@
+ *
+ */
+#pragma once
+
```
Missing `#include "env_validate.h"`

thisiskeithb commented on Marlin/src/pins/ramps/pins_DAGOMA_D6.h:
```diff
@@ -0,0 +1,110 @@
+  #define Y_DIAG_PIN                          41
+  #define Z_DIAG_PIN                          47
+  #define E0_DIAG_PIN                         21
+  #define E1_DIAG_PIN                         20
```
DIAG pins are usually closer to the top of the pins file right above the Limit Switches section / outside of the `HAS_TMC_UART` block.

Sophist-UK committed:
Address @thisiskeithb review comments

thisiskeithb commented on Marlin/src/pins/ramps/pins_DAGOMA_D6.h:
```diff
@@ -0,0 +1,112 @@
+ */
+#pragma once
+
+#include "env_validate.h"
```
I deleted my original comment because `pins_RAMPS.h` handles the environment check/`env_validate.h` include, so this isn't needed.

Sophist-UK replied:
@thisiskeithb No problems.

Sophist-UK committed:
Remove `#include env_validate.h`

thisiskeithb committed:
Add header comments, allow pin overrides

Sophist-UK commented:
Thanks to everyone who helped with this PR - especially @0r31 Orel who has put in much more effort than I did.

sjasonsmith commented:
I observed that @0r31 did earlier work on this and jump in to test. I made sure they were credited as a co-author on the merge.
